curve25519-dalek/src/montgomery.rs

```diff
@@ -394,6 +391,25 @@ impl<'a, 'b> Mul<&'b Scalar> for &'a MontgomeryPoint {
 
         x0.as_affine()
     }
+
+    /// Given `self` \\( = u\_0(P) \\), and a big-endian bit representation of an integer
+    /// \\(n\\), return \\( u\_0(\[n\]P) \\).
+    #[cfg(feature = "hazmat")]
+    fn mul_bits_be(self, bits: impl Iterator<Item = bool>) -> MontgomeryPoint {
+        self.mul_bits_be(bits)
```

Comment:
Is this supposed to be `pub`?

It looks like it calls itself and infinitely recurses? Is it supposed to call `self._mul_bits_be`?

If the intent is for this to be `pub`, I worry about people accidentally discovering it. Feature-gating on `hazmat` alone isn't a great deterrent since other dependencies can activate the feature. Putting it in the `hazmat` module would better prevent accidental discovery/misuse IMO (and free up the `mul_bits_be` name so you don't need a `_`-prefixed version)

At the very least, it could use a more stringent warning label.

Reply:
Oh good points. I agree it'd be best to put it in the `hazmat` module. Will fix.

Reply:
Acutally, this is a question: should this method be feature-gated? There's two main misuses I can think of:

1. If someone used this method instead of `mul_clamped`. This could leak up to 3 secret bits. Not great.
2. If someone used this method instead of `Mul<Scalar> for MontgomeryPoint`. This would be fine, actually. In the worst case, this is equivalent to doing the scalar mul. In the best case, this is _more_ correct than the scalar mul, because the point being multiplied might not be in the prime-order subgroup, and so the scalar mul made no sense anyway.

So the question is if case (1) is likely and worth avoiding. I'm not sure. If someone wanted to multiply by some arbitrary bytestring today, they can already do `Scalar::from_bytes_mod_order(bytes) * point`. If the bytestring is uniform, then this is already equivalent to `mul_bits_be` with 50% probability.

My vote is to let this function be public, and let it have the documentation note I just added.

Reply:
I have a wild suggestion, how do people think about a `RingElement` type that is over the full curve's order? is this too much for that kind of feature?

Reply:
RingElement is overkill imo. We don't define a way to add or multiply two numbers mod 8p, so it'd be a pretty useless ring element. Integer multiplication is broad enough for this use case imo